File: component/biz/Class_biz_role_member_map.pas
unit Class_biz_role_member_map;

interface

uses
  Class_db_role_member_map,
  Class_biz_notifications,
  Class_biz_user,
  system.collections;

type
  TClass_biz_role_member_map = class
  private
    db_role_member_map: TClass_db_role_member_map;
    biz_notifications: TClass_biz_notifications;
    biz_user: TClass_biz_user;
  public
    constructor Create;
    function BePrivilegedToModifyTuple
      (
      has_config_roles_and_matrices: boolean;
      has_assign_roles_to_members: boolean;
      role_natural_text: string
      )
      : boolean;
    procedure Bind
      (
      sort_order: string;
      be_sort_order_descending: boolean;
      target: system.object;
      out crosstab_metadata: arraylist
      );
    procedure BindActuals
      (
      sort_order: string;
      be_sort_order_ascending: boolean;
      target: system.object
      );
    procedure BindHolders
      (
      role_name: string;
      target: system.object;
      sort_order: string;
      be_sort_order_ascending: boolean
      );
    procedure Save
      (
      member_id: string;
      role_id: string;
      be_granted: boolean
      );
  end;

implementation

uses
  kix;

constructor TClass_biz_role_member_map.Create;
begin
  inherited Create;
  db_role_member_map := TClass_db_role_member_map.Create;
  biz_notifications := TClass_biz_notifications.Create;
  biz_user := TClass_biz_user.Create;
end;

function TClass_biz_role_member_map.BePrivilegedToModifyTuple
  (
  has_config_roles_and_matrices: boolean;
  has_assign_roles_to_members: boolean;
  role_natural_text: string
  )
  : boolean;
begin
  BePrivilegedToModifyTuple := has_config_roles_and_matrices
    or ((role_natural_text <> 'Application Administrator') and has_assign_roles_to_members);
end;

procedure TClass_biz_role_member_map.Bind
  (
  sort_order: string;
  be_sort_order_descending: boolean;
  target: system.object;
  out crosstab_metadata: arraylist
  );
begin
  db_role_member_map.Bind
    (
    sort_order,
    be_sort_order_descending,
    target,
    crosstab_metadata
    );
end;

procedure TClass_biz_role_member_map.BindActuals
  (
  sort_order: string;
  be_sort_order_ascending: boolean;
  target: system.object
  );
begin
  db_role_member_map.BindActuals(sort_order,be_sort_order_ascending,target);
end;

procedure TClass_biz_role_member_map.BindHolders
  (
  role_name: string;
  target: system.object;
  sort_order: string;
  be_sort_order_ascending: boolean
  );
begin
  db_role_member_map.BindHolders(role_name,target,sort_order,be_sort_order_ascending);
end;

procedure TClass_biz_role_member_map.Save
  (
  member_id: string;
  role_id: string;
  be_granted: boolean
  );
begin
  db_role_member_map.Save(member_id,role_id,be_granted);
  biz_notifications.IssueForRoleChange(member_id,role_id,be_granted);
end;

end.

File: trunk/p4o/component/biz/Class_biz_role_member_map.cs
using Class_biz_notifications;
using Class_db_role_member_map;
using Class_db_role_member_map_logs;
using System.Collections;

namespace Class_biz_role_member_map
  {
  public class TClass_biz_role_member_map
    {
        private readonly TClass_biz_notifications biz_notifications = null;
        private readonly TClass_db_role_member_map db_role_member_map = null;
        private readonly TClass_db_role_member_map_logs db_role_member_map_logs = null;

        public TClass_biz_role_member_map() : base()
        {
            biz_notifications = new TClass_biz_notifications();
            db_role_member_map = new TClass_db_role_member_map();
            db_role_member_map_logs = new TClass_db_role_member_map_logs();
        }

        public bool BePrivilegedToModifyTuple(bool has_config_roles_and_matrices, bool has_assign_roles_to_members, string role_natural_text)
        {
            bool result;
            result = has_config_roles_and_matrices || ((role_natural_text != "Application Administrator") && has_assign_roles_to_members);
            return result;
        }

        public void Bind(string sort_order, bool be_sort_order_descending, object target, out ArrayList crosstab_metadata)
        {
            db_role_member_map.Bind(sort_order, be_sort_order_descending, target, out crosstab_metadata);
        }

        public void BindActuals(string sort_order, bool be_sort_order_ascending, object target)
        {
            db_role_member_map.BindActuals(sort_order, be_sort_order_ascending, target);
        }

        public void BindHolders(string role_name, object target, string sort_order, bool be_sort_order_ascending)
        {
            db_role_member_map.BindHolders(role_name, target, sort_order, be_sort_order_ascending);
        }

        public void Save(string member_id, string role_id, bool be_granted)
        {
            db_role_member_map.Save(member_id, role_id, be_granted);
            db_role_member_map_logs.Enter
              (
              subject_member_id:member_id,
              be_granted:be_granted,
              role_id:role_id
              );
            biz_notifications.IssueForRoleChange(member_id, role_id, be_granted);
        }

    } // end TClass_biz_role_member_map

}

namespace Class_biz_role_member_map.Units
  {
  public class Class_biz_role_member_map
    {
    } // end Class_biz_role_member_map

}

